File: linkedlist/leetcode/876. Middle of the Linked List.java

// Definition for singly-linked list.
public class ListNode {
  int val;
  ListNode next;
  ListNode(int x) { val = x; }
}
class OwnSolution {
  // O(n) = n+1
  // memeory complexity = O(2n)
  public ListNode middleNodeList(ListNode head) {
    ListNode pointer = head;
    LinkedList<ListNode> list = new LinkedList<>();
    while(pointer != null){
        list.add(pointer);
        pointer = pointer.next;
    }
    return list.get(list.size()%2==0?list.size()/2:(list.size()-1)/2);
  }
  // O(n) = n+1/2n
  // memory complexity = O(n)
  public ListNode middleNodeRevisit(ListNode head) {
    ListNode pointer = head;
    int counter = 0;
    while(pointer != null){
        counter++;
        pointer = pointer.next;
    }
    System.out.println(counter);
    if(counter %2 == 0){
        counter = counter / 2 +1;
    }
    else{
        counter = (counter+1)/2;
    }
    System.out.println(counter);
    ListNode pointerTwo = head;
    for(int i = 0;i<counter-1;i++){
        pointerTwo = pointerTwo.next;
    }
    return pointerTwo;
  }
}

class AnswerSolution{
  
}

File: linkedlist/leetcode/876. Middle of the Linked List.py
# Definition for singly-linked list.
class ListNode:
     def __init__(self, x):
         self.val = x
         self.next = None

class Solution:
    def middleNode(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        list = []
        pointer1 = head
        while(pointer1 != None):
            list.append(pointer1)
            pointer1 = pointer1.next
        if len(list)%2 == 0:
            return list[int(round(len(list)/2))]
        else:
            return list[int(round((len(list)-1)/2))]